lib/mixins/properties-mixin.html

```diff
@@ -0,0 +1,212 @@
+        return props ? Object.keys(props).map(p => {
+          return this.attributeNameForProperty(p);
+        }) : [];
+      }
+
+      /**
+       * Finalizes an element definition, including ensuring any super classes
+       * are also finalized. This includes ensuring property
+       * accessors exist on the element prototype. This method calls
+       * `_finalizeClass` to finalize each constructor in the prototype chain.
+       */
+       static finalize() {
+        if (!this.hasOwnProperty(JSCompiler_renameProperty('__finalized', this))) {
+          const superCtor = superForClass(this);
+          if (superCtor) {
+            superCtor.finalize();
+          }
+          this.__finalized = true;
+          this._finalizeClass();
+        }
+      }
+
+      /**
+       * Finalize an element class. This includes ensuring property
+       * accessors exist on the element prototype. This method is called by
+       * `finalize` and finalizes the class constructor.
+       *
+       * @protected
+       */
+      static _finalizeClass() {
+        const props = ownProperties(this);
+        if (props) {
+          this.createProperties(props);
+        }
+      }
+
+      /**
+       * Returns a memoized version of all properties, including those inherited
+       * from super classes. Properties not in object format are converted to
+       * at lesat {type}.
+       *
+       * @return {Object} Object containing properties for this class
+       * @protected
+       */
+      static get _properties() {
+        if (!this.hasOwnProperty(
+          JSCompiler_renameProperty('__properties', this))) {
+          const superCtor = superForClass(this);
+          this.__properties = Object.assign({},
+            superCtor && superCtor._properties, ownProperties(this));
+        }
+        return this.__properties;
+      }
+
+      /**
+       * Overrides PropertiesChanged method to return type specified in the
+       * static `properties` object for the given property.
+       * @param {string} name Name of property
+       * @return {*} Type to which to deserialize attribute
+       *
+       * @protected
+       */
+      static typeForProperty(name) {
+        const info = this._properties[name];
+        return info && info.type;
+      }
+
+      /**
+       * Overrides default behavior and adds a call to `finalize` which lazily
+       * configures the element's property accessors.
+       * @override
+       */
+      _initializeProperties() {
+        this.constructor.finalize();
+        super._initializeProperties();
+      }
+
+      /**
+       * Called when the element is added to a document.
+       * Calls `_enableProperties` to turn on property system from
+       * `PropertiesChanged`.
+       */
+      connectedCallback() {
+        this._enableProperties();
+      }
+
+      /**
+       * Called when the element is removed from a document
+       */
+      disconnectedCallback() {}
```

Comment:
Is this required to be here?

Reply:
It's there as a minor convenience so users don't have to check before calling `super.disconnectedCallback`

Reply:
The problem with an empty stub is like with my above comment: if the mixin is applied to a superclass with a `disconnectedCallback`, the superclasses method won't be called.

Also, if we do this for `disconnectedCallback` we might want to for `adoptedCallback`?

Reply:
I'm inclined to remove `disconnectedCallback` from this mixin altogether as YAGNI (this mixin is still likely to be used in a higher level base element that handles rendering where that author can make a choice of whether to add a superable stub for convenience).

Reply:
Eh, I can go both ways I guess.  This used to be implemented in `ElementMixin` and just moved up the layering to here (assuming the same reason we had it for `Polymer.Element` will want to be carried forth to future concrete base classes as well).  Having this pattern where users are encouraged to always `super.disconnectedCallback` makes it safer for existing mixins to add a `disconnectedCallback` override at a later point.

Adopted callback is so obscure, I think it would only a waste of bytes. :P